### ISSUE TITLE
Squiggly lines on diagnostic are not rendered correctly

### ISSUE THREAD

**Issue by keyboardDrummer** [dafny-lang/ide-vscode#289]
The squiggly lines on diagnostics are sometimes rendered with gaps in them, instead of as one continuous squiggly line.

<img width="822" alt="image" src="https://user-images.githubusercontent.com/3121201/197217353-18ef405e-52c2-4915-962f-be55c566de28.png">

It can also occur that multiple squiggly lines are rendered across each other:

<img width="956" alt="image" src="https://user-images.githubusercontent.com/3121201/197221651-7b9fb017-7f35-4012-9895-8c843544f6bb.png">

**Comment by MikaelMayer** [comment]
For the first one I think these are two different errors. One is the failing postcondition, the other is the return path.
For the second one, it's different.

**Comment by keyboardDrummer** [comment]
> For the first one I think these are two different errors. One is the failing postcondition, the other is the return path.

But even in the failing postcondition, the part before the `{`, the squiggly lines are not contiguous, they're interrupted at "tall" characters like `(`, `|` and `[`

**Comment by MikaelMayer** [comment]
I close it as won't fix because we have no control over the styling of squiggly bars. The only alternative would be to publish double diagnostics and underline postconditions in red, but I don't think it would result in a good user experience. I am very happy that the two have different colors and slightly different shapes because it helps me distinguish the errors on the well-formedness on the precondition (red) from the postconditions that could not be proven on the body (orange)

**Comment by keyboardDrummer** [comment]
I feel this feature is more for power users. I don't understand what it does, so it just looks broken to me. Do we have an option to turn this it off?

**Comment by MikaelMayer** [comment]
What makes you feel this feature is more for power users?
For a novice user, postconditions feel like tests, and not reporting that a test is failing is not a good user experience. We already iterated on that by adding a red rectangle in front of failing postconditions.

No there is no option to turn it off, I don't see any reason to do so except that it looks broken, which isn't, what you see is just a limitation of VSCode's squiggly line styling. One thing that you might want is to switch highlighting a token instead of a range like it was before and for all other errors.

The orange squiggly bars are highlighting postconditions which are failing for at least one return path. They are not reported as diagnostics, as the action point for the diagnostic is the return path. They are only sub-diagnostics, but equally important in verification.

Let me give you an example of when it can be useful. I start writing a lemma, some requires, and some ensures. I have not yet stepped in the body, as Dafny applies induction directly, so for the first ensures, they serve just as a test to ensure things are as expected. Being able to have squiggly lines on them is priceless, because they behave like assertions if they were inline in the body: I can know whether something is not trivial. And when I'm done writing my ensures, I know exactly which ensures I need to turn into assertions that I will prove using rewriting and the weakest precondition calculus.

**Comment by keyboardDrummer** [comment]
> What makes you feel this feature is more for power users?

Maybe that's not the right term, but for anyone not familiar with what these custom squiggles map to, they can make the IDE look broken. As a VSCode user, I expect the squiggles to map to the diagnostics in the problems view. These squiggles both looks slightly different and 


For a novice user, postconditions feel like tests, and not reporting that a test is failing is not a good user experience. We already iterated on that by adding a red rectangle in front of failing postconditions.
> 
> No there is no option to turn it off, I don't see any reason to do so except that it looks broken, which isn't, what you see is just a limitation of VSCode's squiggly line styling. One thing that you might want is to switch highlighting a token instead of a range like it was before and for all other errors.
> 
> The orange squiggly bars are highlighting postconditions which are failing for at least one return path. They are not reported as diagnostics, as the action point for the diagnostic is the return path. They are only sub-diagnostics, but equally important in verification.
> 
> Let me give you an example of when it can be useful. I start writing a lemma, some requires, and some ensures. I have not yet stepped in the body, as Dafny applies induction directly, so for the first ensures, they serve just as a test to ensure things are as expected. Being able to have squiggly lines on them is priceless, because they behave like assertions if they were inline in the body: I can know whether something is not trivial. And when I'm done writing my ensures, I know exactly which ensures I need to turn into assertions that I will prove using rewriting and the weakest precondition calculus.

**Comment by keyboardDrummer** [comment]
> We can afford this secondary highlight only because the postcondition is part of the same method, hence the same error context. Doing this secondary squiggly line on an unrelated predicate (for postconditions) or function precondition would provide the wrong feeling that verification is not modular.

Makes sense. Thanks!

> I had originally planned to dynamically highlight those remote places, i.e.. mark them as error when the user places the caret at a failing postcondition or a failing return point, but it turns out, often these places are screens apart, and according to the numerous customer feedback I have gathered, people don't want to scroll to a place to see what is wrong exactly. They just want the information to be available on hover.

VSCode already shows related location error messages with a jump to link when you hover over an error diagnostic:
<img width="1321" alt="image" src="https://user-images.githubusercontent.com/3121201/212336208-2a081101-1d43-4971-8d6f-851b8648ad59.png">

What was there to still resolve using hovers?

**Comment by MikaelMayer** [comment]
> VSCode already shows related locatio messages with a jump to link when you hover over an error diagnostic:
image
> What was there to improve using hovers?

Previously, the message was saying "this is the postcondition that might not hold" and there was just a link, which forced people to actually navigate out of the context of the error to see what the failing postcondition was. That was the thing that I improved, which resulted in the message quoting the postcondition code that you now see. Now, users don't need to navigate to another place in the code to see the failing postcondition.

**Comment by keyboardDrummer** [comment]
> > VSCode already shows related location messages with a jump to link when you hover over an error diagnostic:
> > image
> > What was there to improve using hovers?
> 
> Previously, the message was saying "this is the postcondition that might not hold" and there was just a link, which forced people to actually navigate out of the context of the error to see what the failing postcondition was. That was the thing that I improved, which resulted in the message quoting the postcondition code that you now see. Now, users don't need to navigate to another place in the code to see the failing postcondition.

Makes sense. I've created a related ticket in the VSCode repo: https://github.com/microsoft/vscode/issues/171401

I think I meant to ask about LSP Hover information though. Sometimes errors are also shown in there, such as here:

<img width="578" alt="image" src="https://user-images.githubusercontent.com/3121201/212642913-1facb800-b34d-40d2-973b-5e6eb847e7b4.png">

Why is that done?

**Comment by MikaelMayer** [comment]
> I think I meant to ask about LSP Hover information though. Sometimes errors are also shown in there, such as here:
> 
> <img alt="image" width="578" src="https://user-images.githubusercontent.com/3121201/212642913-1facb800-b34d-40d2-973b-5e6eb847e7b4.png">
> 
> Why is that done?

* To put a link on "Error" that redirect to the page of Verification debugging, which is helpful to fix verification errors
* To display the assertion number (useful to determine which assertion is taking too long if :vcs_split_on_every_assert is used),
* To display the assertion count, to determine whether it might be worth splitting an assertion batch..
* To display the resource count of the assertion batch, and if it is too big, to display a warning with a link to verification debugging when verification is slow.

> Have we considered well enough, flipping the ranges of the diagnostic and its related location for postcondition failures? Maybe that would make the related location squiggles obsolete?

Yes, not in this thread, but in offline conversations, we determined that the error's main position should be where the assertion is physically implicitly present. Errors should be where users are typically expected to modify something to fix them.
When postconditions fails, there are typically two fixes:  either fix the specification, or fix the code. We found that the latter one is much much more prevalent.
